images.go

```diff
@@ -637,20 +686,31 @@ func (r *imageStore) SetBigData(id, key string, data []byte) error {
 			image.BigDataNames = append(image.BigDataNames, key)
 			save = true
 		}
-		if key == ImageDigestBigDataKey {
-			if oldDigest != "" && oldDigest != newDigest && oldDigest != image.Digest {
-				// remove the image from the list of images in the digest-based
-				// index which corresponds to the old digest for this item, unless
-				// it's also the hard-coded digest
-				if list, ok := r.bydigest[oldDigest]; ok {
-					prunedList := imageSliceWithoutValue(list, image)
-					if len(prunedList) == 0 {
-						delete(r.bydigest, oldDigest)
-					} else {
-						r.bydigest[oldDigest] = prunedList
-					}
+		for _, oldDigest := range image.Digests {
+			// remove the image from the list of images in the digest-based index
+			if list, ok := r.bydigest[oldDigest]; ok {
+				prunedList := imageSliceWithoutValue(list, image)
+				if len(prunedList) == 0 {
+					delete(r.bydigest, oldDigest)
+				} else {
+					r.bydigest[oldDigest] = prunedList
 				}
 			}
+		}
+		digestList := make([]digest.Digest, 0, len(image.BigDataDigests)+1)
+		if image.Digest != "" {
+			digestList = append(digestList, image.Digest)
+		}
+		for itemName, itemDigest := range image.BigDataDigests {
+			if bigDataNameIsManifest(itemName) {
+				digestList = append(digestList, itemDigest)
+			}
+		}
+		image.Digests, err = recomputeDigests(digestList)
+		if err != nil {
+			return errors.Wrapf(err, "error loading recomputing image naming information for %s", image.ID)
+		}
+		for _, newDigest := range image.Digests {
 			// add the image to the list of images in the digest-based index which
 			// corresponds to the new digest for this item, unless it's already there
 			list := r.bydigest[newDigest]
```

Comment:
Above, all entries of (old) `image.Digests` have been removed from `r.bydigest`; doesn’t that mean that `len == len(imageSliceWithoutValue)` below will always be true?

Reply:
In practice it _should_, but uniqueness of manifests among images isn't something we're enforcing at this level.

Reply:
Ah, right; I realized at some point that IDs are not necessarily config digests but did not notice how it applies to this.